MHOUSE-6810: Support for declaring post-task actions

timbokopter opened the pull request:
**Context:** `DependsOn()` allows tasks to build on the actions performed by other tasks. This is especially useful when multiple tasks require the same or similar setup steps, e.g. integration/e2e test tasks that need to build the same binaries, generate artifacts or prepare the filesystem before executing individual tests.

**The Problem:** Tasks may involve similarly redundant/repetitive actions after executing to e.g. revert setup steps, clean up temporary artifacts or perform other post processing actions, which can currently only be done by the task executor itself. This adds possibly redundant overhead and requires careful handling of errors to ensure post-task actions are not inadvertently skipped. Deferring the work to e.g. make generated artifacts available to other tasks via `DependsOn()` before being removed is also not possible.

**The Solution:** This PR adds the builder method `Finally(names ...string)` to optionally declare one or more tasks that should be run after the task itself has executed, regardless of its error status.

---

**Design Notes**
- `Finally()` tasks only execute if the underlying task was invoked, but regardless of whether it succeeded or failed
- If multiple tasks are run, their `Finally()` tasks will be run in reverse sort order, i.e. `A.DependsOn(B)` means `B` is executed *before* `A`, but its `Finally()` tasks are run *after* `A` and its `Finally()`
- `Finally()` can be used on any task except pure aggregate tasks which don't have an executor
- Parameters for `Finally()` can be any task(s) that **do not**
	- use `Finally()` themselves
	- have required parameters
	- depend on tasks matching either of the above
- Even if specified by multiple tasks, each `Finally()` task is only executed **once**. While generally executed in the order they were specified, a `Finally()` task use by a higher-level task may change the effective order for a lower-level one.
	- Ex: `T1.Finally("F1", "F2", "F3")`, `T2.DependsOn("T1").Finally("F2")`
	- Running `T2` executes `T1 -> T2 [finalize] F2 -> F1 -> F3`

craiggwilson commented on task/builder.go:
```diff
@@ -87,6 +87,12 @@ func (b *Builder) Hide() *Builder {
 }
 
+// Finally declares other tasks that should run after this one.
+func (b *Builder) Finally(names ...string) *Builder {
```
Let's go with `Defer` here!

craiggwilson commented on task/run.go:
```diff
@@ -93,6 +97,31 @@ func Run(registry *Registry, arguments []string) error {
+		writer.SetPrefix(prefix)
+		startTime := time.Now()
+		for _, task := range finallyTasks {
+			if task.Executor() != nil {
```
```suggestion
			if executor := task.Executor(); executor != nil {
```

craiggwilson commented on task/run.go:
```diff
@@ -93,6 +97,31 @@ func Run(registry *Registry, arguments []string) error {
+		startTime := time.Now()
+		for _, task := range finallyTasks {
+			if task.Executor() != nil {
+				if err := task.Executor()(ctx); err != nil {
```
```suggestion
				if err := executor(ctx); err != nil {
```

craiggwilson commented on task/sort.go:
```diff
@@ -38,7 +47,12 @@ func buildGraph(allTasks []Task, requiredTaskNames []string) ([]*graphNode, erro
+			if err := validateFinallyTasks(allTasksMap, finallyTaskStates, task.Finally()); err != nil {
+				return nil, err
+			}
+			// toposort can
```
unfinished comment?

timbokopter replied:
I thought this comment sufficiently explained the situation, but sure, happy to elaborate

craiggwilson commented on task/task.go:
```diff
@@ -11,6 +11,7 @@ type Task interface {
 	Executor() Executor
 	Hidden() bool
 	Name() string
+	Finally() []string
```
`Defer` here as well.

craiggwilson reviewed (commented):
Just a couple of comments... looks good though. Thanks!

timbokopter commented:
Addressed the comments (hopefully). Also included an additional change to properly pass arguments to deferred tasks, which now also allows tasks with required arguments.
